lib/quicly.c

```diff
@@ -1548,6 +1561,36 @@ static void do_free_pn_space(struct st_quicly_pn_space_t *space)
     free(space);
 }
 
+static int change_outside_reorder_window(quicly_ranges_t *ranges, uint64_t *smallest_unreported_missing, uint64_t received_pn,
+                                         uint8_t reordering_threshold)
+{
+    if (ranges->num_ranges == 0 || reordering_threshold == 0)
+        return 0;
+
+    uint64_t prev_smallest_unreported_missing = *smallest_unreported_missing;
+    uint64_t largest_pn = ranges->ranges[ranges->num_ranges - 1].end - 1;
+
+    if (received_pn == prev_smallest_unreported_missing) {
+        if (received_pn == largest_pn) {
+            // fast path
+            *smallest_unreported_missing = largest_pn + 1;
+        } else {
+            *smallest_unreported_missing = quicly_ranges_next_missing(ranges, received_pn + 1);
+        }
+    }
+
+    if (largest_pn < reordering_threshold)
+        return 0;
+
+    uint64_t largest_pn_outside_reorder_window = largest_pn - (uint64_t)reordering_threshold;
+
+    if (*smallest_unreported_missing <= largest_pn_outside_reorder_window)
+        *smallest_unreported_missing = quicly_ranges_next_missing(ranges, largest_pn_outside_reorder_window + 1);
+
+    return (prev_smallest_unreported_missing <= largest_pn_outside_reorder_window) ||
+           received_pn <= largest_pn_outside_reorder_window;
+}
```

Comment:
Consider the following sequence:
1. `reordering_threshold` is 2, `ranges` is [0..2], `*smallest_unreported_missing` is 3.
2. quicly_receive(PN=5) is called. It elicits an immediate ACK and sets `*smallest_unreported_missing` to 4.
3. As an immediate ack is elicited, quicly_send() is called immediately, reporting both 3 _and 4_ are missing.

Based on my read of the PR (sorry if I'm incorrect), `smallest_unreported_missing` becomes 4, but it should be set to 6, because all packets up to 5 are reported.

I wonder what we should be doing is:
* at any moment we decide to ACK immediately), set `*smallest_unreported_missing` to UNT64_MAX, and then
* in `send_ack`, set to largest_pn + 1.